2022-2023/Laboratories/2022-2023-advopsys-lab3-dtrace.ipynb

```diff
@@ -0,0 +1,187 @@
+    "    else:\n",
+    "        print(\"## Failed with the exit code {}\".format(output_dtrace.returncode))\n",
+    "        \n",
+    "    # Explicitly free DTrace resources.\n",
+    "    # Python's Garbage Collector would free DTrace resources when\n",
+    "    # dtrace_thread is reassigned, e.g. when the cell is reexecuted.\n",
+    "    # This could be confusing when analysing kernel from a terminal\n",
+    "    # and the notebook at the same time.\n",
+    "    del dtrace_thread"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d84dea26-685a-4eb5-8526-28d9d6118289",
+   "metadata": {},
+   "source": [
+    "## Collect TCP segment details and system-call counts\n",
+    "\n",
+    "In order to collect both information on TCP segments (as mentioned in Advanced Operating Systems: Lab 3 – TCP, General Information) and system-call counts (as in Advanced Operating Systems: Lab 1 - Getting Started with Kernel Tracing), we define two actions: one that prints to output details on a TCP segment, and one that aggregates system-call counts.\n",
+    "\n",
+    "Our `out` function called `tcp_out` parses bytes from one output line at the time and must decode information printed to output as opposed to the `walker` function `tcp_walker` that receives a list of keys.\n",
+    "\n",
+    "Note that the `tcp_script` D script is an example D script that should be extended with appropriate predicates to only trace information relevant to our benchmark."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "89ec029c-e13b-48f5-b4c8-03fbddd75fc0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tcp_script = \"\"\"\n",
+    "fbt::tcp_do_segment:entry\n",
+    "{\n",
+    "    printf(\"%u %u %s\",\n",
+    "        (unsigned int)args[1]->th_seq,\n",
+    "        (unsigned int)args[1]->th_ack,\n",
+    "        tcp_state_string[args[3]->t_state]);\n",
+    "}\n",
+    "\n",
+    "syscall:::entry\n",
+    "/execname == \"ipc-benchmark\"/\n",
+    "{\n",
+    "    @syscalls[probefunc] = count();\n",
+    "}\n",
+    "\"\"\"\n",
+    "\n",
+    "from collections import defaultdict\n",
+    "syscall_count_values = defaultdict(int)\n",
+    "tcp_segments = []\n",
+    "\n",
+    "def tcp_walker(action, identifier, keys, value):\n",
+    "    \"\"\"\n",
+    "    action -- a type of action (sum, avg, ...)\n",
+    "    identifier -- the id\n",
+    "    keys -- list of keys\n",
+    "    value -- the value\n",
+    "    \"\"\"\n",
+    "    syscall_count_values[keys[0]] += value\n",
+    "\n",
+    "def tcp_out(value):\n",
+    "    \"\"\"\n",
+    "    value -- the value, of the bytes type.\n",
+    "    \"\"\"\n",
+    "    value = value.decode('ascii').split(' ')\n",
+    "    tcp_segments.append({'seq': int(value[0]), 'ack': int(value[1]), 'state': value[2]})\n",
+    "\n",
+    "dtrace_synchronous(tcp_script, tcp_walker, tcp_out, \"ipc/ipc-benchmark -j -v -i tcp -b 64 -t 64 2thread\")\n",
```

Comment:
FYI I use `-b 64 -t 64` because we would get a lot of output otherwise. Students are expected to process data rather than print all lines.